checker/src/main/java/org/checkerframework/checker/determinism/DeterminismAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,990 @@
+    }
+
+    /** @return true if {@code method} is a main method */
+    public static boolean isMainMethod(ExecutableElement method) {
+        if (method.getReturnType().getKind() == TypeKind.VOID
+                && method.getSimpleName().contentEquals("main")
+                && method.getParameters().size() == 1
+                && method.getParameters().get(0).asType().getKind() == TypeKind.ARRAY) {
+            ArrayType arrayType = (ArrayType) method.getParameters().get(0).asType();
+            if (TypesUtils.isString(arrayType.getComponentType())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Adds implicit annotation for main method formal parameter ({@code @Det}) and default
+     * annotations for the component types of other array formal parameters ({@code ...[@PolyDet]}).
+     *
+     * <p>Note: The annotation on an array type defaults to {@code @PolyDet[]} and this defaulting
+     * is handled by declarative mechanism.
+     *
+     * <p>Example: Consider the following code:
+     *
+     * <pre><code>
+     * &nbsp; void testArr(int[] a) {
+     * &nbsp;   ...
+     * &nbsp; }
+     * </code></pre>
+     *
+     * This method {@code addComputedTypeAnnotations} annotates the component type of parameter
+     * {@code int[] a} as {@code @PolyDet int[] a}.
+     *
+     * <p>Note: Even though {@code visitExecutable} and {@code addComputedTypeAnnotations} have the
+     * same logic for adding defaults to parameter types, the code structure is different. This is
+     * because the argument to {@code visitExecutable} is an {@code AnnotatedExecutableType} which
+     * represents the type of a method, constructor, or initializer, and the argument to {@code
+     * addComputedTypeAnnotations} is any {@code Element}.
+     */
+    @Override
+    public void addComputedTypeAnnotations(Element elt, AnnotatedTypeMirror type) {
+        if (elt.getKind() == ElementKind.PARAMETER) {
+            if (elt.getEnclosingElement().getKind() == ElementKind.METHOD) {
+                ExecutableElement method = (ExecutableElement) elt.getEnclosingElement();
+                if (isMainMethod(method)) {
+                    if (!type.getAnnotations().isEmpty() && !type.hasAnnotation(DET)) {
+                        checker.report(
+                                Result.failure(
+                                        "invalid.annotation.on.parameter",
+                                        type.getAnnotationInHierarchy(NONDET)),
+                                elt);
+                    }
+                    type.addMissingAnnotations(Collections.singleton(DET));
+                }
+
+                defaultArrayComponentTypeAsPolyDet(type);
+            }
+        }
+        super.addComputedTypeAnnotations(elt, type);
+    }
+
+    /** @return true if {@code tm} is Set or a subtype of Set */
+    private boolean isSet(TypeMirror tm) {
```

Comment:
This comment applies to all of these methods that check if something is a subtype of a certain class.

`AnnotatedTypeMirror#getUnderlyingType` is called before most of the calls to these methods.  This complicates the code and makes it harder to read.  Instead the parameters of these methods should be AnnotatedTypeMirrors.

The body of all these methods is identical except for which element is compared.   Instead of duplicating the code, call a method `isSubClassOf` that takes an AnnotatedTypeMirror and a TypeMirror.